### PR TITLE
Container tab listing

### PR DESCRIPTION
**Pull request by bakulf**
The UI is not completed but the logic and all the API between child and parent process is done.

### CODE REVIEW

**Comment by bakulf** [comment]
This covers issues #31, #22 (partially) and #10 partially

File: index.js [hunk 1]
```diff
@@ -219,6 +223,110 @@ let ContainerService = {
     });
   },
 
+  getTabs(args) {
+    return new Promise((resolve, reject) => {
+      if (!("userContextId" in args)) {
+        reject(null);
```
Comment (groovecoder) [inline index.js]: Need a reason error message here. e.g., `reject("getTabs must be called with userContextId argument.")`

File: index.js [hunk 2]
```diff
@@ -219,6 +223,110 @@ let ContainerService = {
...
+        return;
+      }
+
+      let list = [];
+      let promises = [];
+
+      for (let tab of tabs) {
+        let xulTab = viewFor(tab);
+        let userContextId = parseInt(xulTab.getAttribute("usercontextid") || 0, 10);
```
Comment (groovecoder) [inline index.js]: I notice this line (and the preceding line) is repeated a lot. Can it be refactored into a helper/utility function to reduce code duplication?

File: index.js [hunk 3]
```diff
@@ -219,6 +223,110 @@ let ContainerService = {
...
+      Promise.all(promises).then(() => {
+        resolve(list);
+      });
+    });
+  },
+
+  showTab(args) {
+    return new Promise((resolve, reject) => {
+      if (!("tabId" in args)) {
+        reject(null);
```
Comment (groovecoder) [inline index.js]: Need a reason error message here. e.g., reject("showTab must be called with tabId argument.")

File: index.js [hunk 4]
```diff
@@ -219,6 +223,110 @@ let ContainerService = {
...
+        return;
+      }
+
+      let list = [];
+      let promises = [];
+
+      for (let tab of tabs) {
+        let xulTab = viewFor(tab);
+        let userContextId = parseInt(xulTab.getAttribute("usercontextid") || 0, 10);
+        if (args.userContextId == userContextId) {
```
Comment (groovecoder) [inline index.js]: Should check without type coercion? (`===`)

File: index.js [hunk 5]
```diff
@@ -219,6 +223,110 @@ let ContainerService = {
...
+
+  showTab(args) {
+    return new Promise((resolve, reject) => {
+      if (!("tabId" in args)) {
+        reject(null);
+        return;
+      }
+
+      for (let tab of tabs) {
+        if (tab.id == args.tabId) {
```
Comment (groovecoder) [inline index.js]: Should check without type coercion? (`===`)

File: index.js [hunk 6]
```diff
@@ -219,6 +223,110 @@ let ContainerService = {
...
+          tab.activate();
+          break;
+        }
+      }
+
+      resolve(null);
+    });
+  },
+
+  moveTabsToWindow(args) {
```
Comment (groovecoder) [inline index.js]: OMG ❤️ this feature.

File: index.js [hunk 7]
```diff
@@ -219,6 +223,110 @@ let ContainerService = {
...
+    return new Promise((resolve, reject) => {
+      if (!("userContextId" in args)) {
+        reject(null);
```
Comment (groovecoder) [inline index.js]: Need a reason error message here. e.g., `reject("moveTabsToWindow must be called with userContextId argument.")`

File: index.js [hunk 8]
```diff
@@ -219,6 +223,110 @@ let ContainerService = {
...
+
+      // Let"s create a list of the tabs.
+      let list = [];
+      for (let tab of tabs) {
+        let xulTab = viewFor(tab);
+        let userContextId = parseInt(xulTab.getAttribute("usercontextid") || 0, 10);
+        if (args.userContextId == userContextId) {
+          list.push(xulTab);
+        }
+      }
```
Comment (groovecoder) [inline index.js]: Almost the same code as in `getTabs` - maybe a helper/utility function like:
```
function getTabList(userContextId) {
  let list = [];
  ...
  return list;
}
```
?

File: index.js [hunk 9]
```diff
@@ -219,6 +223,110 @@ let ContainerService = {
...
+      // Nothing to do
+      if (list.length == 0) {
+        resolve(null);
+        return;
+      }
+
+      windows.browserWindows.open({
+        url: "about:blank",
+        onOpen: window => {
+          let browserWindow = viewFor(window);
```
Comment (groovecoder) [inline index.js]: Can this var name be explicit that it's the **new** browser window? `newBrowserWindow`?

File: webextension/js/popup.js [hunk 10]
```diff
@@ -2,33 +2,64 @@
...
-    });
-  })
-  // Let"s update the icon.
-  .then((identity) => {
-    let hideorshowIcon = document.querySelector(`#uci-${identity.userContextId}-hideorshow-icon`);
-    if (!identity.hasHiddenTabs && !identity.hasOpenTabs) {
-      hideorshowIcon.style.display = "none";
-    } else {
-      hideorshowIcon.style.display = "";
+function showContainerPanel(identity) {
```
Comment (groovecoder) [inline webextension/js/popup.js]: Can this be explicit? `showContainerTabsPanel`?

File: webextension/js/popup.js [hunk 11]
```diff
@@ -2,33 +2,64 @@
...
 
-    hideorshowIcon.src = hasHiddenTabs ? CONTAINER_HIDE_SRC : CONTAINER_UNHIDE_SRC;
+    document.getElementById("container-info-table").appendChild(fragment);
+  })
 
-    // The new identity is returned.
-    return identity;
+  // Finally we are ready to show the panel.
+  .then(() => {
+    // FIXME: the animation...
```
Comment (groovecoder) [inline webextension/js/popup.js]: Note: I've got a `showPanel` utility function coming that we can use here.

File: webextension/js/popup.js [hunk 12]
```diff
@@ -139,3 +145,33 @@ document.querySelector("#sort-containers-link").addEventListener("click", () =>
     window.close();
   });
 });
+
+document.querySelector("#close-container-info-panel").addEventListener("click", () => {
+  // TODO: animation
```
Comment (groovecoder) [inline webextension/js/popup.js]: Again, we can use `showPanel` here when we merge up with my latest.

File: webextension/popup.html [hunk 13]
```diff
@@ -48,6 +48,29 @@
       </div>
     </div>
   </div>
+  <div class="hide panel container-info-panel" id="container-info-panel">
```
Comment (groovecoder) [inline webextension/popup.html]: I'll plan to re-write this to match UI spec after we merge up.

File: index.js (hunk 4 above)
Comment (pdehaan) [inline index.js]: I added this ESLint rule in #66, so you may need to rebase. 👏

File: webextension/js/popup.js (hunk 11 above)
Comment (bakulf) [inline webextension/js/popup.js]: This will be a follow up

File: webextension/popup.html (hunk 13 above)
Comment (bakulf) [inline webextension/popup.html]: \o/

**Comment by bakulf** [comment]
Thanks for your review. I totally agree with all your comments. Can you take a look again?

File: index.js [hunk 14]
```diff
@@ -144,14 +145,28 @@ let ContainerService = {
     };
   },
 
+  _getUserContextIdFromTab(tab) {
```
Comment (groovecoder) [inline index.js]: :+1:

File: index.js [hunk 15]
```diff
@@ -144,14 +145,28 @@ let ContainerService = {
...
+    return parseInt(viewFor(tab).getAttribute("usercontextid") || 0, 10);
+  },
+
+  _getTabList(userContextId) {
```
Comment (groovecoder) [inline index.js]: 👍

File: index.js [hunk 16]
```diff
@@ -52,24 +56,21 @@ let ContainerService = {
 
     // It can happen that this jsm is loaded after the opening a container tab.
     for (let tab of tabs) {
-      let xulTab = viewFor(tab);
-      let userContextId = parseInt(xulTab.getAttribute("usercontextid") || 0, 10);
+      let userContextId = this._getUserContextIdFromTab(tab);
```
Comment (jonathanKingston) [inline index.js]: Nit: const

File: index.js [hunk 17]
```diff
@@ -52,24 +56,21 @@ let ContainerService = {
...
       if (userContextId) {
         ++this._identitiesState[userContextId].openTabs;
       }
     }
 
     tabs.on("open", tab => {
-      let xulTab = viewFor(tab);
-      let userContextId = parseInt(xulTab.getAttribute("usercontextid") || 0, 10);
+      let userContextId = this._getUserContextIdFromTab(tab);
```
Comment (jonathanKingston) [inline index.js]: Nit: const

File: index.js [hunk 18]
```diff
@@ -52,24 +56,21 @@ let ContainerService = {
...
       if (userContextId) {
         ++this._identitiesState[userContextId].openTabs;
       }
     });
 
     tabs.on("close", tab => {
-      let xulTab = viewFor(tab);
-      let userContextId = parseInt(xulTab.getAttribute("usercontextid") || 0, 10);
+      let userContextId = this._getUserContextIdFromTab(tab);
```
Comment (jonathanKingston) [inline index.js]: Nit: const

File: index.js [hunk 19]
```diff
@@ -144,14 +145,28 @@ let ContainerService = {
     };
   },
 
+  _getUserContextIdFromTab(tab) {
+    return parseInt(viewFor(tab).getAttribute("usercontextid") || 0, 10);
+  },
+
+  _getTabList(userContextId) {
+    let list = [];
+    for (let tab of tabs) {
```
Comment (jonathanKingston) [inline index.js]: Nit: are we doing const in loops?

File: index.js [hunk 20]
```diff
@@ -144,14 +145,28 @@ let ContainerService = {
     };
   },
 
+  _getUserContextIdFromTab(tab) {
+    return parseInt(viewFor(tab).getAttribute("usercontextid") || 0, 10);
+  },
+
+  _getTabList(userContextId) {
+    let list = [];
```
Comment (jonathanKingston) [inline index.js]: Nit: const

File: index.js [hunk 21]
```diff
@@ -144,14 +145,28 @@ let ContainerService = {
...
+
   // Tabs management
 
   hideTabs(args) {
     return new Promise(resolve => {
       for (let tab of tabs) {
-        let xulTab = viewFor(tab);
-        let userContextId = parseInt(xulTab.getAttribute("usercontextid") || 0, 10);
-
+        let userContextId = this._getUserContextIdFromTab(tab);
```
Comment (jonathanKingston) [inline index.js]: Nit: const

File: index.js [hunk 22]
```diff
@@ -217,6 +232,96 @@ let ContainerService = {
...
   },
 
+  getTabs(args) {
+    return new Promise((resolve, reject) => {
+      if (!("userContextId" in args)) {
+        reject("getTabs must be called with userContextId argument.");
+        return;
+      }
+
+      let list = this._getTabList(args.userContextId);
```
Comment (jonathanKingston) [inline index.js]: Nit: const

File: index.js [hunk 23]
```diff
@@ -217,6 +232,96 @@ let ContainerService = {
...
+
+  moveTabsToWindow(args) {
+    return new Promise((resolve, reject) => {
+      if (!("userContextId" in args)) {
+        reject("moveTabsToWindow must be called with userContextId argument.");
+        return;
+      }
+
+      // Let"s create a list of the tabs.
+      let list = this._getTabList(args.userContextId);
```
Comment (jonathanKingston) [inline index.js]: Nit: const

File: index.js [hunk 24]
```diff
@@ -217,6 +232,96 @@ let ContainerService = {
...
+      // Nothing to do
+      if (list.length === 0) {
+        resolve(null);
+        return;
+      }
+
+      windows.browserWindows.open({
+        url: "about:blank",
+        onOpen: window => {
+          let newBrowserWindow = viewFor(window);
```
Comment (jonathanKingston) [inline index.js]: Nit: const

File: index.js [hunk 25]
```diff
@@ -217,6 +232,96 @@ let ContainerService = {
...
+
+          // Let's move the tab to the new window.
+          for (let tab of list) {
+            let newTab = newBrowserWindow.gBrowser.addTab("about:blank");
```
Comment (jonathanKingston) [inline index.js]: Nit: const

File: webextension/js/popup.js [hunk 26]
```diff
@@ -2,33 +2,64 @@
...
+  icon.setAttribute("data-identity-icon", identity.image);
+  icon.setAttribute("data-identity-color", identity.color);
+
+  // Show or not the has-tabs section.
+  for (let trHasTabs of document.getElementsByClassName("container-info-has-tabs")) {
+    trHasTabs.style.display = !identity.hasHiddenTabs && !identity.hasOpenTabs ? "none" : "";
+    trHasTabs.setAttribute("data-user-context-id", identity.userContextId);
+  }
+
+  let hideShowIcon = document.getElementById("container-info-hideorshow-icon");
```
Comment (jonathanKingston) [inline webextension/js/popup.js]: Nit: const and others in this file... sorrrrry

File: webextension/js/popup.js [hunk 27]
```diff
@@ -139,3 +145,33 @@ document.querySelector("#sort-containers-link").addEventListener("click", () =>
     window.close();
   });
 });
+
+document.querySelector("#close-container-info-panel").addEventListener("click", () => {
+  // TODO: animation
+  document.getElementById("container-info-panel").classList.add("hide");
```
Comment (jonathanKingston) [inline webextension/js/popup.js]: Can we use hidden or will this be animated?

**Pull request merged**